crates/zizmor/src/github_api.rs

```diff
@@ -282,16 +378,61 @@ impl Client {
 
     #[instrument(skip(self))]
     #[tokio::main]
-    pub(crate) async fn list_tags(&self, owner: &str, repo: &str) -> Result<Vec<Tag>, ClientError> {
-        self.paginate(&format!("repos/{owner}/{repo}/tags"))
+    pub(crate) async fn list_branches(
+        &self,
+        owner: &str,
+        repo: &str,
+    ) -> Result<Vec<Branch>, ClientError> {
+        self.list_branches_internal(owner, repo).await
+    }
+
+    async fn list_tags_internal(&self, owner: &str, repo: &str) -> Result<Vec<Tag>, ClientError> {
+        self.list_refs(owner, repo)
             .await
+            .map(|v| {
+                let mut tags: Vec<_> = v
+                    .iter()
+                    .filter_map(|r| {
+                        r.name.strip_prefix("refs/tags/").map(|name| Tag {
+                            name: name.to_string(),
+                            commit: Commit {
+                                sha: r.oid.to_string(),
+                            },
+                        })
+                    })
+                    .collect();
+
+                // Tags may point to a commit or an annotation.
+                // If we have an annotation, a tag suffixed with `^{}` holds the commit.
+                // See: https://www.kernel.org/pub/software/scm/git/docs/gitrevisions.html
+                let annotated_tags: HashSet<_> = tags
+                    .iter()
+                    .filter_map(|tag| tag.name.strip_suffix("^{}").map(|n| n.to_string()))
+                    .collect();
+                tags.retain_mut(|tag| {
+                    if let Some(stripped_name) = tag.name.strip_suffix("^{}") {
+                        tag.name = stripped_name.to_string();
+                        true
+                    } else {
+                        !annotated_tags.contains(&tag.name)
+                    }
+                });
```

Comment:
I haven't fully thought through this yet, but I think there might be an edge case here -- tags can point to other tags (not just commits), so to fully flatten a tag to its underlying commit we probably need to recurse here.

(OTOH, we have tests for `ref-version-mismatch` that test this exact situation and they're passing, so I guess this works as is? I'm tired.)